src/java.base/macosx/native/libjli/java_md_macosx.m

```diff
@@ -60,115 +60,52 @@
 #define LD_LIBRARY_PATH "DYLD_FALLBACK_LIBRARY_PATH"
 
 /*
- * If a processor / os combination has the ability to run binaries of
- * two data models and cohabitation of jre/jdk bits with both data
- * models is supported, then DUAL_MODE is defined. MacOSX is a hybrid
- * system in that, the universal library can contain all types of libraries
- * 32/64 and client/server, thus the spawn is capable of linking with the
- * appropriate library as requested.
+ * Following is the high level flow of the launcher code residing in the common java.c and this
+ * macosx specific java_md_macosx file:
  *
- * Notes:
- * 1. VM. DUAL_MODE is disabled, and not supported, however, it is left here in
- *    for experimentation and perhaps enable it in the future.
- * 2. At the time of this writing, the universal library contains only
- *    a server 64-bit server JVM.
- * 3. "-client" command line option is supported merely as a command line flag,
- *    for, compatibility reasons, however, a server VM will be launched.
- */
-
-/*
- * Flowchart of launcher execs and options processing on unix
+ *  - JLI_Launch function, which is the entry point to the launcher, calls CreateExecutionEnvironment
+ *
+ *  - CreateExecutionEnvironment does the following (not necessarily in that order):
+ *      - determines the relevant JVM type that needs to be ultimately created
+ *      - determines the path and asserts the presence of libjava and relevant libjvm library
+ *      - removes any JVM selection options from the arguments that were passed to the launcher
+ *
+ *  - CreateExecutionEnvironment then creates a new thread, within the same process, to launch the application's main()
+ *    Java method and parks the current thread, on which CreateExecutionEnvironment was invoked, in Apple's Cocoa event
+ *    loop. Before doing so, CreateExecutionEnvironment maintains a state flag to keep note that a new thread
+ *    has been spawned.
+ *
+ *  - The newly created thread (in which the application's main() method will ultimately run) starts right from the
+ *    beginning of the current process' main function, which effectively means that JLI_Launch is re-invoked on this
+ *    new thread and the same above sequence of code flow repeats again. During this "recursive" call, when at the
+ *    point of creating a new thread in CreateExecutionEnvironment, the CreateExecutionEnvironment will check for
+ *    the state flag to see if a new thread has already been spawned and upon noticing that it has, it will skip
+ *    spawning any more threads and will return back from CreateExecutionEnvironment.
```

Comment:
My understanding was that this recursive invocation was only needed when switching modes/models. I don't see why it would be needed now.

Reply:
In the context of the 2 platforms - macosx and unix, the recursive invocation still continues to happen.

For macosx, the `CreateExecutionEnvironment` unconditionally, through an internal `MacOSXStartup` function, spawns a new thread (`pthread_create`). That thread is passed a function pointer pointing to the current executable/process' `main(...)` function (and thus made to execute afresh). That then triggers this recursion where `JLI_Launch` is called again and that then calls into `CreateExecutionEnvironment` all the way to `MacOSXStartup` function which has the necessary knowledge not to spawn another thread the second time.

For unix, in the `CreateExecutionEnvironment` function, there's a specific piece of code which determines whether or not `LD_LIBRARY_PATH` environment variable needs to be set or updated before loading the JVM. This decision of whether or not `LD_LIBRARY_PATH` needs to be updated or set is done in an internal function called `RequiresSetenv`. There are several rules which determine whether or not to set/update that environment variable. Rules like - is musl libc in use; or is the runtime environment AIX; or if the `LD_LIBRARY_PATH` is currently set, then whether the value in the environment variable matches some well known JVM library path patterns (like `lib/client`, `lib/server`). If any of these rules determine that the `LD_LIBRARY_PATH` needs to be set/updated, then the `CreateExecutionEnvironment` function will do the necessary updates to the environment variable value and and exec() the current process with that new value. This then triggers the recursion all the way from the `JLI_Launch` back into this `CreateExecutionEnvironment` (which has the necessary knowledge not to exec() once more).

Reply:
> For macosx, the CreateExecutionEnvironment unconditionally, through an internal MacOSXStartup function, spawns a new thread (pthread_create).

I forgot to note that, the reason for spawning this new thread is explained as a code comment (unrelated to the changes in this PR) on the `MacOSXStartup` function and it says:

```
/*
* Mac OS X mandates that the GUI event loop run on very first thread of
* an application. This requires that we re-call Java's main() on a new
* thread, reserving the 'main' thread for Cocoa.
*/

```

Reply:
That aside, the launcher always tries to spawn a new thread because bad stuff can happen on some OS if we try to make the process primordial thread a JavaThread (e.g. you can't enforce stack size limits).

But I remain unclear as to why the new thread has to "start again" rather than continuing with the launch process- it is even called ContinueInNewThread.

Reply:
> But I remain unclear as to why the new thread has to "start again" rather than continuing with the launch process

That's unclear to me as well. As a separate task, I do plan to look into the historical changes in this area to understand better and see if we can improve this part to not restart afresh right from the beginning and instead continue the launch process from a specific point.